packages/mana/basevector.go

```diff
@@ -133,3 +135,59 @@ func (bmv *BaseManaVector) GetWeightedMana(nodeID identity.ID, weight float64) (
 func (bmv *BaseManaVector) GetMana(nodeID identity.ID) (float64, error) {
 	return bmv.GetWeightedMana(nodeID, 0.5)
 }
+
+// ForEach iterates over the vector and calls the provided callback.
+func (bmv *BaseManaVector) ForEach(callback func(ID identity.ID, bm *BaseMana) bool) {
+	for nodeID, baseMana := range bmv.vector {
+		if !callback(nodeID, baseMana) {
+			return
+		}
+	}
+}
+
+//GetManaMap return mana perception of the node.
+func (bmv *BaseManaVector) GetManaMap() Map {
+	res := make(map[identity.ID]float64)
+	for ID := range bmv.vector {
+		mana, _ := bmv.GetMana(ID)
+		res[ID] = mana
+	}
+	return res
+}
+
+// GetHighestManaNodes return the n highest mana nodes in descending order.
+// It also updates the mana values for each node.
+func (bmv *BaseManaVector) GetHighestManaNodes(n int) []NodeIDMana {
+	var res []NodeIDMana
+	for ID := range bmv.vector {
+		mana, _ := bmv.GetMana(ID)
+		res = append(res, NodeIDMana{
+			ID:   ID,
+			Mana: mana,
+		})
+	}
+
+	sort.Slice(res[:], func(i, j int) bool {
+		return res[i].Mana > res[j].Mana
+	})
+
+	if n <= len(res) {
+		return res[:n]
+	}
+	return res[:]
+}
+
+// SetMana sets the base mana for a node
+func (bmv *BaseManaVector) SetMana(nodeID identity.ID, bm *BaseMana) {
+	bmv.vector[nodeID] = bm
+	_ = bmv.Update(nodeID, time.Now())
+}
+
+// NodeIDMana represents a node and its mana value.
+type NodeIDMana struct {
```

Comment:
let's just call it `Node`. It has an ID and a Mana value associated with it.

Reply:
yeah